File: src/service.ts
/*
 *   Copyright (c) 2026. caoccao.com Sam Cao
 *   All rights reserved.

 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at

 *   http://www.apache.org/licenses/LICENSE-2.0

 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

import { invoke } from "@tauri-apps/api/core";
import type {
  About,
  BetterMediaInfoStatus,
  Config,
  MergeSnapshot,
  MediaMetadata,
  MkvToolNixStatus,
  UpdateCheckResult,
} from "./protocol";

export async function getAbout(): Promise<About> {
  return await invoke<About>("get_about");
}

export async function getConfig(): Promise<Config> {
  return await invoke<Config>("get_config");
}

export async function setConfig(config: Config): Promise<Config> {
  return await invoke<Config>("set_config", { config });
}

export async function getMediaFiles(paths: string[]): Promise<string[]> {
  return await invoke<string[]>("get_media_files", { paths });
}

export async function getLaunchArgs(): Promise<string[]> {
  return await invoke<string[]>("get_launch_args");
}

export async function isMkvtoolnixFound(
  path: string,
  checkRunning: boolean = false,
): Promise<MkvToolNixStatus> {
  return await invoke<MkvToolNixStatus>("is_mkvtoolnix_found", {
    path,
    checkRunning,
  });
}

export async function getMediaMetadata(file: string): Promise<MediaMetadata> {
  return await invoke<MediaMetadata>("get_media_metadata", { file });
}

export async function enqueueMerge(
  file: string,
  args: string[],
): Promise<void> {
  return await invoke<void>("enqueue_merge", { file, args });
}

export async function cancelMerge(file: string): Promise<void> {
  return await invoke<void>("cancel_merge", { file });
}

export async function getMergeStatus(): Promise<MergeSnapshot> {
  return await invoke<MergeSnapshot>("get_merge_status");
}

export async function checkOutputPathWritable(path: string): Promise<boolean> {
  return await invoke<boolean>("check_output_path_writable", { path });
}

export async function ensureOutputPath(path: string): Promise<void> {
  return await invoke<void>("ensure_output_path", { path });
}

/** `<outputDir>/<sourceStem>.mkv`, with " (1)", " (2)", … appended when a file
 *  with that name already exists, so merging never overwrites. */
export async function resolveMergeOutputPath(
  outputDir: string,
  sourceFile: string,
): Promise<string> {
  return await invoke<string>("resolve_merge_output_path", {
    outputDir,
    sourceFile,
  });
}

export async function detectBetterMediaInfo(
  path: string,
  checkRunning: boolean = false,
): Promise<BetterMediaInfoStatus> {
  return await invoke<BetterMediaInfoStatus>("detect_better_media_info", {
    path,
    checkRunning,
  });
}

export async function launchBetterMediaInfo(paths: string[]): Promise<void> {
  return await invoke<void>("launch_better_media_info", { paths });
}

export async function getUpdateResult(): Promise<UpdateCheckResult | null> {
  return await invoke<UpdateCheckResult | null>("get_update_result");
}

export async function skipVersion(version: string): Promise<void> {
  return await invoke<void>("skip_version", { version });
}

File: src-tauri/src/media_metadata/mpeg_ts/descriptors/service.rs
/*
 *   Copyright (c) 2026. caoccao.com Sam Cao
 *   All rights reserved.

 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at

 *   http://www.apache.org/licenses/LICENSE-2.0

 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

//! DVB Service descriptor (tag 0x48) — ETSI EN 300 468 §6.2.33.
//!
//! Body layout:
//!
//! ```text
//! u8 service_type
//! u8 service_provider_name_length
//! [service_provider_name_length bytes of DVB string]
//! u8 service_name_length
//! [service_name_length bytes of DVB string]
//! ```
//!
//! DVB strings can carry an encoding prefix byte (0x01..=0x1F) — we strip it
//! and treat the remainder as Latin-1 / UTF-8 best-effort.

pub fn decode(body: &[u8]) -> Option<String> {
    if body.len() < 2 {
        return None;
    }
    // skip service_type (1) + provider_name
    let provider_len = body[1] as usize;
    let after_provider = 2 + provider_len;
    if after_provider >= body.len() {
        return None;
    }
    let service_name_len = body[after_provider] as usize;
    let name_start = after_provider + 1;
    let name_end = name_start + service_name_len;
    if name_end > body.len() {
        return None;
    }
    let raw = &body[name_start..name_end];
    Some(decode_dvb_string(raw))
}

fn decode_dvb_string(bytes: &[u8]) -> String {
    if bytes.is_empty() {
        return String::new();
    }
    let body = if matches!(bytes[0], 0x01..=0x1F) {
        &bytes[1..]
    } else {
        bytes
    };
    // Try UTF-8 first; fall back to Latin-1 lossy.
    match std::str::from_utf8(body) {
        Ok(s) => s.to_string(),
        Err(_) => body.iter().map(|b| *b as char).collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn body(service_type: u8, provider: &[u8], service_name: &[u8]) -> Vec<u8> {
        let mut v = Vec::new();
        v.push(service_type);
        v.push(provider.len() as u8);
        v.extend_from_slice(provider);
        v.push(service_name.len() as u8);
        v.extend_from_slice(service_name);
        v
    }

    #[test]
    fn extracts_service_name_utf8() {
        let b = body(0x01, b"BBC", b"BBC One");
        assert_eq!(decode(&b).as_deref(), Some("BBC One"));
    }

    #[test]
    fn strips_dvb_charset_prefix() {
        let b = body(0x01, b"BBC", &[0x05, b'B', b'B', b'C']); // 0x05 = ISO-8859-9
        assert_eq!(decode(&b).as_deref(), Some("BBC"));
    }

    #[test]
    fn rejects_truncated_body() {
        assert!(decode(&[0x01]).is_none());
        assert!(decode(&[0x01, 5, b'a', b'b']).is_none());
    }

    #[test]
    fn returns_empty_when_service_name_zero_length() {
        let b = body(0x01, b"BBC", b"");
        assert_eq!(decode(&b).as_deref(), Some(""));
    }
}
